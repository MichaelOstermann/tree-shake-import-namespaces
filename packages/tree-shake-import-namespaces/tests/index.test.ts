import { describe, expect, it } from "vitest"
import { defaultOptions, expectSnapshot } from "./helpers"

describe("babel-plugin-tree-shake-imports", () => {
    it("Should tree-shake named imports", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar;
        `, {
            resolveImport(data) {
                expect(data).toEqual({
                    filePath: "source.tsx",
                    importAlias: "_bar",
                    importName: "Foo",
                    importPath: "foo",
                    localName: "Foo",
                    propertyName: "bar",
                    scope: new Set(["Foo"]),
                })
                return defaultOptions.resolveImport(data)
            },
        })
    })

    it("Should tree-shake aliased imports", () => {
        expectSnapshot(`
            import { Foo as Bar } from "foo";
            Bar.bar;
        `, {
            resolveImport(data) {
                expect(data).toEqual({
                    filePath: "source.tsx",
                    importAlias: "_bar",
                    importName: "Foo",
                    importPath: "foo",
                    localName: "Bar",
                    propertyName: "bar",
                    scope: new Set(["Bar"]),
                })
                return defaultOptions.resolveImport(data)
            },
        })
    })

    it("Should tree-shake default imports", () => {
        expectSnapshot(`
            import Foo from "foo";
            Foo.bar;
        `, {
            resolveImport(data) {
                expect(data).toEqual({
                    filePath: "source.tsx",
                    importAlias: "_bar",
                    importName: undefined,
                    importPath: "foo",
                    localName: "Foo",
                    propertyName: "bar",
                    scope: new Set(["Foo"]),
                })
                return defaultOptions.resolveImport(data)
            },
        })
    })

    it("Should tree-shake wildcard imports", () => {
        expectSnapshot(`
            import * as Foo from "foo";
            Foo.bar;
        `, {
            resolveImport(data) {
                expect(data).toEqual({
                    filePath: "source.tsx",
                    importAlias: "_bar",
                    importName: "*",
                    importPath: "foo",
                    localName: "Foo",
                    propertyName: "bar",
                    scope: new Set(["Foo"]),
                })
                return defaultOptions.resolveImport(data)
            },
        })
    })

    it("Should transform JSX", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            const Component = (
                <Foo.Bar>
                </Foo.Bar>
            );
        `, {
            resolveImport(data) {
                expect(data).toEqual({
                    filePath: "source.tsx",
                    importAlias: "_Bar",
                    importName: "Foo",
                    importPath: "foo",
                    localName: "Foo",
                    propertyName: "Bar",
                    scope: new Set(["Foo", "Component"]),
                })
                return defaultOptions.resolveImport(data)
            },
        })
    })

    it("Should tree-shake multiple references", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar;
            Foo.bar;
            Foo.baz;
            Foo.baz;
        `)
    })

    it("Should not tree-shake nested imports by default", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar.baz;
        `)
    })

    it("Should tree-shake nested imports when requested", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar.baz;
        `, {
            ...defaultOptions,
            nested: true,
        })
    })

    it("Should not tree-shake when not desired", () => {
        expectSnapshot(`
            import { A } from "a";
            import { B } from "b";
            A.a;
            B.c;
            B.d;
        `, {
            resolveImport(data) {
                if (data.importName === "B" && data.propertyName === "d") return
                return defaultOptions.resolveImport(data)
            },
        })
    })

    it("Should create unique identifiers", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar;
            const bar = true;
            const _bar = true;
        `)
    })

    it("Should create unique identifiers across scopes", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar;
            const example = function (bar) {
                Foo.bar;
            };
            const example2 = function (_bar) {
                Foo.bar;  
            };
        `)
    })

    it("Should create unique identifiers across namespaces", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            import { Bar } from "bar";
            Foo.baz;
            Bar.baz;
        `)
    })

    it("Should skip irrelevant references", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            function example() {
                const Foo = {};
                return Foo.bar;
            };
        `)
    })

    it("Should skip fishy properties", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar;
            Foo["bar"];
        `)
    })

    it("Should keep unused imports", () => {
        expectSnapshot(`
            import { Foo, Bar } from "foo";
            Foo.bar;
        `)
    })
})
