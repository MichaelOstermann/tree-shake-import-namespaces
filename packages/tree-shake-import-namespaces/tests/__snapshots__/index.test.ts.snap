// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`babel-plugin-tree-shake-imports > Should create unique identifiers 1`] = `
"import { bar as _bar1 } from "foo/bar";
_bar1;
const bar = true;
const _bar = true;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { Foo } from \\"foo\\";\\nFoo.bar;\\nconst bar = true;\\nconst _bar = true;"
  ],
  "names": [],
  "mappings": "AAAA;AACA,KAAO;AACP,KAAK,CAAC,GAAG,CAAC,CAAC,CAAC,IAAI;AAChB,KAAK,CAAC,IAAI,CAAC,CAAC,CAAC,IAAI"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should create unique identifiers across namespaces 1`] = `
"import { baz as _baz } from "foo/baz";
import { baz as _baz1 } from "bar/baz";
_baz;
_baz1;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { Foo } from \\"foo\\";\\nimport { Bar } from \\"bar\\";\\nFoo.baz;\\nBar.baz;"
  ],
  "names": [],
  "mappings": "AAAA;AACA;AACA,IAAO;AACP,KAAO"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should create unique identifiers across scopes 1`] = `
"import { bar as _bar1 } from "foo/bar";
_bar1;
const example = function (bar) {
    _bar1;
};
const example2 = function (_bar) {
    _bar1;  
};

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { Foo } from \\"foo\\";\\nFoo.bar;\\nconst example = function (bar) {\\n    Foo.bar;\\n};\\nconst example2 = function (_bar) {\\n    Foo.bar;  \\n};"
  ],
  "names": [],
  "mappings": "AAAA;AACA,KAAO;AACP,KAAK,CAAC,OAAO,CAAC,CAAC,CAAC,QAAQ,CAAC,CAAC,GAAG,CAAC,CAAC;AAC/B,CAAC,CAAC,CAAC,CAAC,KAAO;AACX,CAAC;AACD,KAAK,CAAC,QAAQ,CAAC,CAAC,CAAC,QAAQ,CAAC,CAAC,IAAI,CAAC,CAAC;AACjC,CAAC,CAAC,CAAC,CAAC,KAAO,CAAC,CAAC;AACb,CAAC"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should keep unused imports 1`] = `
"import {Bar} from "foo";
import { bar as _bar } from "foo/bar";

_bar;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { Foo, Bar } from \\"foo\\";\\nFoo.bar;"
  ],
  "names": [],
  "mappings": "AAAA;;;AACA,IAAO"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should not tree-shake nested imports by default 1`] = `
"import { bar as _bar } from "foo/bar";
_bar.baz;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { Foo } from \\"foo\\";\\nFoo.bar.baz;"
  ],
  "names": [],
  "mappings": "AAAA;AACA,IAAO,CAAC,GAAG"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should not tree-shake when not desired 1`] = `
"import { a as _a } from "a/a";
import { B } from "b";
_a;
B.c;
B.d;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { A } from \\"a\\";\\nimport { B } from \\"b\\";\\nA.a;\\nB.c;\\nB.d;"
  ],
  "names": [],
  "mappings": "AAAA;AACA,MAAM,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,IAAI,CAAC,CAAC,CAAC,CAAC;AACrB,EAAG;AACH,CAAC,CAAC,CAAC;AACH,CAAC,CAAC,CAAC"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should skip fishy properties 1`] = `""`;

exports[`babel-plugin-tree-shake-imports > Should skip irrelevant references 1`] = `""`;

exports[`babel-plugin-tree-shake-imports > Should transform JSX 1`] = `
"import { Bar as _Bar } from "foo/Bar";
const Component = (
    <_Bar>
    </_Bar>
);

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { Foo } from \\"foo\\";\\nconst Component = (\\n    <Foo.Bar>\\n    </Foo.Bar>\\n);"
  ],
  "names": [],
  "mappings": "AAAA;AACA,KAAK,CAAC,SAAS,CAAC,CAAC,CAAC;AAClB,CAAC,CAAC,CAAC,CAAC,CAAC,IAAO;AACZ,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,IAAO;AACb,CAAC"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake aliased imports 1`] = `
"import { bar as _bar } from "foo/bar";
_bar;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { Foo as Bar } from \\"foo\\";\\nBar.bar;"
  ],
  "names": [],
  "mappings": "AAAA;AACA,IAAO"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake default imports 1`] = `
"import { bar as _bar } from "foo/bar";
_bar;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import Foo from \\"foo\\";\\nFoo.bar;"
  ],
  "names": [],
  "mappings": "AAAA;AACA,IAAO"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake multiple references 1`] = `
"import { bar as _bar } from "foo/bar";
import { baz as _baz } from "foo/baz";
_bar;
_bar;
_baz;
_baz;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { Foo } from \\"foo\\";\\nFoo.bar;\\nFoo.bar;\\nFoo.baz;\\nFoo.baz;"
  ],
  "names": [],
  "mappings": "AAAA;AAAA;AACA,IAAO;AACP,IAAO;AACP,IAAO;AACP,IAAO"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake named imports 1`] = `
"import { bar as _bar } from "foo/bar";
_bar;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { Foo } from \\"foo\\";\\nFoo.bar;"
  ],
  "names": [],
  "mappings": "AAAA;AACA,IAAO"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake nested imports when requested 1`] = `
"import { baz as _baz } from "foo/bar/baz";
_baz;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import { Foo } from \\"foo\\";\\nFoo.bar.baz;"
  ],
  "names": [],
  "mappings": "AAAA;AACA,IAAW"
}"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake wildcard imports 1`] = `
"import { bar as _bar } from "foo/bar";
_bar;

{
  "version": 3,
  "sources": [
    "source.tsx"
  ],
  "sourcesContent": [
    "import * as Foo from \\"foo\\";\\nFoo.bar;"
  ],
  "names": [],
  "mappings": "AAAA;AACA,IAAO"
}"
`;
